### PR TITLE
Add maximum roughness cutoff to SSR to improve performance

### PR DESCRIPTION
**Pull request by Calinou**
`master` version of https://github.com/godotengine/godot/pull/56815.

In a test scene with mixed rough and non-rough materials, this saves upwards of 0.15 ms of GPU time with very little visual artifacting (GTX 1080, 2560×1440). This improves performance the most in scenes that have mostly rough materials. The optimization works regardless of the SSR roughness quality selected in the Project Settings.

The roughness cutoff was empirically chosen – it seems to strike a good balance between improving performance and avoiding obvious visual artifacts. The roughness cutoff could be exposed to a project setting if needed, but the Params push constant is already full (128 bytes).

**Testing project:** [test_ssr_performance.zip](https://github.com/godotengine/godot/files/10200602/test_ssr_performance.zip)

## Preview

**Some more comparisons:** [Trying to find a good roughness cutoff](https://imgsli.com/OTA2NTU/1/2), [High-quality settings](https://imgsli.com/OTA2NTk)

#### Before

2022-01-15_01 28 23_no_cutoff

#### After

2022-01-15_01 27 40_cutoff_above_075

#### Before

2022-01-15_01 29 26_no_cutoff

#### After

2022-01-15_01 30 13_cutoff_above_075

### CODE REVIEW

**Comment by fire** [comment]
I made a pr in the same area of roughness https://github.com/godotengine/godot/pull/56822.

File: servers/rendering/renderer_rd/shaders/screen_space_reflection.glsl [hunk 1]
```diff
@@ -74,6 +74,14 @@ void main() {
 
 	vec4 normal_roughness = imageLoad(source_normal_roughness, ssC);
 	vec3 normal = normal_roughness.xyz * 2.0 - 1.0;
+	float roughness = normal_roughness.w;
+
+	if (roughness > 0.75) {
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/screen_space_reflection.glsl:79]: Can be reduced to 0.6 to match https://github.com/godotengine/godot/pull/69828
```suggestion
	if (roughness > 0.6) {
```

File: servers/rendering/renderer_rd/shaders/screen_space_reflection.glsl [hunk 2]
```diff
@@ -74,6 +74,14 @@ void main() {
...
+		// Do not compute SSR for rough materials to improve performance at the cost of
+		// subtle artifacting.
+		return;
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/shaders/screen_space_reflection.glsl:82]: We still need to write something to the output or else we will get artifacts in the subsequent passes. Just treat this as a no collision situation and write out black
```suggestion
		// Do not compute SSR for rough materials to improve performance at the cost of
		// subtle artifacting.
#ifdef MODE_ROUGH
		imageStore(blur_radius_image, ssC, vec4(0.0));
#endif
		imageStore(ssr_image, ssC, vec4(0.0));
		return;
```

**Review by clayjohn — COMMENTED** [review]
With the changes listed here this could be merged after https://github.com/godotengine/godot/pull/69828

**Comment by Calinou** [comment]
Benchmark with the new roughness cutoff of 0.6 in 3840×2160 on a AMD Radeon RX 6900 XT (Fedora 36, RADV):

Before | After (this PR)
-|-
479 FPS (2.08 mspf) | 551 FPS (1.81 mspf)

**Review by clayjohn — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Comment by Calinou** [comment]
Despite the comment in the shader, I noticed the roughness cutoff in `master` doesn't actually match https://github.com/godotengine/godot/pull/69828, since it's `0.6` instead of `0.7`:

https://github.com/godotengine/godot/blob/950f4c3b41acef0403618dd6169be90530dea5da/servers/rendering/renderer_rd/shaders/effects/screen_space_reflection.glsl#L71-L72

https://github.com/godotengine/godot/blob/950f4c3b41acef0403618dd6169be90530dea5da/servers/rendering/renderer_rd/shaders/effects/screen_space_reflection.glsl#L226-L228

In practice, this appears to have no visual impact, but restoring the "correct" cutoff value of `0.7` makes performance slightly worse (+0.03 mspf in 3840×2160 on a GeForce RTX 4090):

### `0.6` (`master`)

Screenshot_20230627_110801 webp

### `0.7` (proposed change)

Screenshot_20230627_110859 webp

What should we do? For reference, I've made https://github.com/godotengine/godot/pull/56815 use the "correct" cutoff value of `0.7`, but maybe we should make it use `0.6`.

**Pull request merged**